source/llassetgen/include/llassetgen/packing/internal/Common.h

```diff
@@ -0,0 +1,158 @@
+         */
+        template <class InputIter>
+        size_t rectReserveCount(InputIter begin, InputIter end) {
+            using IterCategory = typename std::iterator_traits<InputIter>::iterator_category;
+            return rectReserveCountInternal(begin, end, IterCategory{});
+        }
+
+        template <class Iter>
+        constexpr int checkIteratorType() {
+            using IterTraits = typename std::iterator_traits<Iter>;
+            using IterRefType = typename IterTraits::reference;
+            using IterCategory = typename IterTraits::iterator_category;
+            static_assert(std::is_assignable<Vec2<PackingSizeType>, IterRefType>::value,
+                          "Packing input elements must be assignable to Vec2<PackingSizeType>");
+            static_assert(std::is_base_of<std::input_iterator_tag, IterCategory>::value,
+                          "Packing input iterator must be an InputIterator");
+
+            // C++11 doesn't allow void as the return type of constexpr functions
+            return 0;
+        }
+
+        /**
+         * Return the smallest integer x so that i <= 2^x.
+         */
+        LLASSETGEN_API unsigned int ceilLog2(uint64_t num);
+
+        /**
+         * Find the smallest atlas size that might fit the given rectangles.
+         *
+         * The returned sizes will always be powers of two, and will be as close
+         * to square as possible.
+         */
+        LLASSETGEN_API Vec2<PackingSizeType> predictAtlasSize(const Packing& packing);
+
+        /**
+         * Given rect sizes, create a packing with rectangles with those sizes.
+         */
+        template <class InputIter>
+        Packing initPacking(InputIter sizesBegin, InputIter sizesEnd) {
+            checkIteratorType<InputIter>();
+
+            Packing packing;
+            packing.rects.reserve(rectReserveCount(sizesBegin, sizesEnd));
+            std::transform(sizesBegin, sizesEnd, std::back_inserter(packing.rects),
+                           [](const Vec2<PackingSizeType>& rectSize) {
+                               return Rect<PackingSizeType>{{0, 0}, rectSize};
+                           });
+            return packing;
+        }
+
+        template <class Packer>
+        bool packAll(Packing& packing, Packer& packer) {
+            std::vector<size_t> indices(packing.rects.size());
+            std::iota(indices.begin(), indices.end(), 0);
+            std::sort(indices.begin(), indices.end(), [&packing](size_t i1, size_t i2) {
+                return Packer::inputSortingComparator(packing.rects[i1], packing.rects[i2]);
+            });
+            return std::all_of(std::begin(indices), std::end(indices),
+                               [&](size_t i) { return packer.pack(packing.rects[i]); });
+        }
+
+        /**
+         * Create a flexible size packing from given rectangle sizes.
+         *
+         * @tparam Packer
+         *   Packer class. Must provide the following methods:
```

Comment:
Why not add the required methods to `BasePacker` and assert that `Packer` is a subclass of `BasePacker`?

Reply:
For either run-time member and static functions as well as compile-time functions and variables there are different idioms to ensure subclass interface availability.
* run-time member function: abstract/virtual member functions (subclassing required)
* run-time static functions: immediate/deferred nested member lookup / concepts (subclassing suitable but not required)
* compile-time functions: nested member lookup / concepts (subclassing suitable but not required)

Reply:
I would like to avoid virtual methods, especially because I would have to make one static method non-static. I could move the documentation to `BasePacker`, but that class basically just the common members/methods from `MaxRectsPacker` and `ShelfPacker` factored out into a class, I don't know if that is a better place.

In general though, this packing class interface is completely internal and only needed when defining a new packing algorithm. Also the compiler errors are relatively clear when a method is missing, so I don't think we really need to check at compile time that the methods exist, just describe somewhere what they should do.

Compiler error example:
```
/Users/david/Documents/Uni/adcpp/source/llassetgen/include/llassetgen/packing/internal/Common.h:86:32: error: no member named 'inputSortingComparator' in 'llassetgen::internal::MaxRectsPacker'
                return Packer::inputSortingComparator(packing.rects[i1], packing.rects[i2]);
```

Reply:
Fair enough. Why do you want to avoid virtual methods?